### PR TITLE
Address PROTOTYPE comments in features/ref-fields

### PR DESCRIPTION
**Pull request by cston**
Address PROTOTYPE comments.

Also fixes https://github.com/dotnet/roslyn/issues/62085.

Proposal: [low-level-struct-improvements.md](https://github.com/dotnet/csharplang/blob/main/proposals/low-level-struct-improvements.md)
Test plan: https://github.com/dotnet/roslyn/issues/59194

### CODE REVIEW

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md [hunk 1]
```diff
@@ -89,6 +89,59 @@ Note: The break will also apply to C# 10 and earlier when .NET 7 ships, but is
...
+
+```csharp
+static ref T ReturnRefParamByRef<T>(ref T t)
+{
+    t = default;
+    return ref t; // ok
+}
+```
+
+## Method ref struct return escape analysis depends on ref escape of ref arguments
```
Comment (AlekseyTs) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:116]: No workaround for these?

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs [hunk 2]
```diff
@@ -1378,6 +1378,7 @@ internal enum ReservedAttributes
             NativeIntegerAttribute = 1 << 9,
             CaseSensitiveExtensionAttribute = 1 << 10,
             RequiredMemberAttribute = 1 << 11,
+            LifetimeAnnotationAttribute = 1 << 12,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs:1381]: >LifetimeAnnotationAttribute = 1 << 12, [](http://example.com/codeflow?start=12&length=38)

I assume we don't want to allow this attribute anywhere in source. In that case, it is probably better to add an explicit check to ```Symbol.DecodeWellKnownAttribute```. More robust and less changes.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Symbol.cs:1381]: Especially that it looks like we don't actually perform the check. I think we shouldn't add an infrastructure it, if it is not used. It should be done, if at all, when #62124 will be addressed.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNodeExtensions.cs [hunk 3]
```diff
@@ -227,29 +228,29 @@ internal static RefKind GetRefKind(this TypeSyntax syntax)
...
-            {
-                syntax = ((RefTypeSyntax)syntax).Type;
-            }
-
-            return syntax;
+            return SkipRef(syntax, out refKind, allowScoped: true, diagnostics: null);
         }
 
-        internal static TypeSyntax SkipRef(this TypeSyntax syntax, out RefKind refKind)
+        internal static TypeSyntax SkipRef(this TypeSyntax syntax, out RefKind refKind, bool allowScoped, BindingDiagnosticBag? diagnostics)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNodeExtensions.cs:236]: >diagnostics [](http://example.com/codeflow?start=128&length=11)

Should we assert that if, ```allowScoped == false```,  diagnostics must be not null?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs [hunk 4]
```diff
@@ -326,7 +326,6 @@ private void EnsureSignatureIsLoaded()
                 _packedFlags.SetRefKind(refKind);
                 _packedFlags.SetIsVolatile(customModifiersArray.Any(static m => !m.IsOptional && ((CSharpCustomModifier)m).ModifierSymbol.SpecialType == SpecialType.System_Runtime_CompilerServices_IsVolatile));
 
-                // PROTOTYPE: `fixed ref` field use should be disallowed.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs:329]: >// PROTOTYPE: `fixed ref` field use should be disallowed. [](http://example.com/codeflow?start=16&length=57)

I didn't see any special handling around that for PE.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs:329]: >// PROTOTYPE: `fixed ref` field use should be disallowed. [](http://example.com/codeflow?start=16&length=57)

Are we tracking similar issue for volatile or already handling that?

**Comment by AlekseyTs** [comment]
>             // PROTOTYPE: `fixed ref` field use should be disallowed.

The comment doesn't appear to be addressed. There probably should be an error about the field being unsupported.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:454 in 9ace414. [](commit_id = 9ace414bae241d48d3500a979220d420d9b5a7a7, deletion_comment = True)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 10)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 5]
```diff
@@ -6603,13 +6880,51 @@ public void ParameterScope_07()
...
+    static void F6(scoped scoped in R r) { }
+    static void F7(out scoped scoped R r) { r = default; }
+    static void F8(scoped out scoped R r) { r = default; }
+    static void F9(scoped scoped out R r) { r = default; }
 }";
             var comp = CreateCompilation(source);
-            // PROTOTYPE: Should report duplicate modifiers.
-            comp.VerifyEmitDiagnostics();
+            comp.VerifyEmitDiagnostics(
+                // (4,27): error CS1107: A parameter can only have one 'scoped' modifier
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:6896]: I found this slightly surprising. I thought we were going to make 'scoped' as a type name an error like we did for 'required' as a type name.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 6]
```diff
@@ -6791,8 +7161,48 @@ struct S1
...
+        public void ExtensionThisScope()
+        {
+            var source =
+@"ref struct R<T> { }
+static class Extensions
+{
+    static void F1(this scoped R<object> r) { }
+    static void F2<T>(scoped this R<T> r) { }
+    static void F3<T>(this scoped ref T t) where T : struct { }
+    static void F4(this ref scoped R<int> r) { }
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:7174]: consider also checking a `this` parameter with no scoped modifiers as a baseline.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs [hunk 7]
```diff
@@ -259,7 +219,25 @@ internal static DeclarationModifiers MakeModifiers(NamedTypeSymbol containingTyp
                 containingType.CheckUnsafeModifier(result, errorLocation, diagnostics);
             }
 
+            if (isRefField)
+            {
+                reportBadMemberFlagIfAny(result, DeclarationModifiers.Static, diagnostics, errorLocation);
```
[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:224]: Yes, I extracted the local function for use with existing cases and to add diagnostics for ref fields. The behavior of the existing cases should be unchanged.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:6896]: We're not reporting an error for `scoped` as a  type name yet.

**Comment by AlekseyTs** [comment]
>   .field public int32& modreq([mscorlib]System.Runtime.CompilerServices.IsVolatile) F

Are we testing scenario with the modifier before the ```&```?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:473 in 972408c. [](commit_id = 972408ca550a6ee62d9c0faed9a6d2dd14780548, deletion_comment = False)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 13), assuming CI is passing

**Pull request merged**